music/src/core/visualizer.ts

```diff
@@ -437,7 +444,8 @@ export class PianoRollSVGVisualizer extends BaseVisualizer {
   }
 
   private drawNote(
-      x: number, y: number, w: number, h: number, fill: string, index: number) {
+      x: number, y: number, w: number, h: number, fill: string,
+      dataAttributes: {[attr: string]: any}) { // tslint:disable-line:no-any
```

Comment:
wahhhh I know this is annoying, but could you please make a type for this (maybe `SVGNoteDataAttributes`?). You can look at the definition of `VisualizerConfig` for what that looks like.

Reply:
So like this?
```typescript
interface SVGNoteDataAttributes {
  index: number,
  instrument: number,
  ...
}
```
But can we then iterate over the attributes [here](https://github.com/tensorflow/magenta-js/pull/399/files/c343792292d8701c19bb6605af30d738c728db92#diff-cde90f4f7fada5c5c2fb344599c23cbeR461)? If not, then I think it would be better to just pass the actual note in.

Also, it might better to use `rect.dataset.isDrum` instead of `rect.setAttribute('is-drum')` ([`dataset`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/dataset) takes care of translating the camel-cased names to the dash-separated ones).

Reply:
Oh no, i meant more like:
(...fill: string, dataAttributes: []: SVGNoteAttribute)
interface SVGNoteDataAttributes {
  key: String,
  value: String
}

And then you can iterate over the array.

And yeah, definitely use dataset. my bad, didn't see that the first time around.

Reply:
OK, got it. Am I allowed to do `value: any` (suppressing `no-any`)? Otherwise I guess I have to `.toString()` everything (and first check if it's defined).

Sorry, this is my first time using TypeScript. :sweat_smile:

Reply:
I mean, I would have to do something like this for every attribute:
```typescript
if (note.instrument !== undefined) {
  dataAttributes.push({key: 'instrument', value: `${note.instrument}`});
} 
```
(Because in the visualizer demo, it does happen that `note.instrument === undefined`.)

Reply:
you can push them as undefined and then not set them, no? you already check if they're undefined before setting them

Reply:
Typescript enforces you to cast them to string before you put them in the dataset anyway (so `foo.dataset.batman = String(someNumber)`, so I don't think you need to do the type `any` in there

Reply:
Yes, but this way `value` cannot be `string`, because I'm passing in numbers and booleans (and `undefined`s)...

Reply:
Do you want me to push it like that so you can have a look?

Reply:
If I turn it into a string before passing it into `drawNote`, I need to check for `undefined`, otherwise it will get turned into `"undefined"`, won't it? And if want to do it inside `drawNote`, I need the `any` to pass it in.

Reply:
And I seem to be able to do `rect.dataset[key] = value` without TypeScript complaining. :thinking: